src/prof.c

```diff
@@ -1705,11 +1710,74 @@ prof_getpid(void) {
 static bool
 prof_dump_maps(bool propagate_err) {
 	bool ret;
-	int mfd;
+	int mfd = -1;
 
 	cassert(config_prof);
 #ifdef __FreeBSD__
-	mfd = prof_open_maps("/proc/curproc/map");
+	int mib[4];
+	size_t sz;
+	char *buf;
+
+	mib[0] = CTL_KERN;
+	mib[1] = KERN_PROC;
+	mib[2] = KERN_PROC_VMMAP;
+	mib[3] = prof_getpid();
+
+	if (prof_dump_write(propagate_err, "\nMAPPED_LIBRARIES:\n") &&
+	    propagate_err) {
+		ret = false;
+		goto label_return;
+	}
+
+	if (sysctl(mib, 4, NULL, &sz, NULL, 0) == -1) {
+		ret = true;
+		goto label_return;
+	}
+
+	buf = mmap(NULL, sz, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANON, -1, 0);
+
+	if (buf == MAP_FAILED) {
+		ret = true;
+		goto label_return;
+	}
+
+	// Size for kinfo_vmentry
+	sz = sz * 4 / 3;
+	if (sysctl(mib, 4, buf, &sz, NULL, 0) == -1) {
+		ret = true;
+		goto label_return;
+	}
+
+	char *low = buf;
+	char *high = buf + sz;
+
+	while (low < high) {
+		char map[PATH_MAX+68];
+		struct kinfo_vmentry *e = (struct kinfo_vmentry *)low;
+		size_t esz = e->kve_structsize;
+
+		if (esz == 0) {
+			ret = true;
+			goto label_return;
+		}
+
+		if (e->kve_path[0] == '\0') {
+			goto loop_cont;
+		}
+
+		snprintf(map, sizeof(map), "%.12lu-%.12lu %c%c%c %s\n",
+			 e->kve_start,
+			 e->kve_end,
+			 (e->kve_protection & KVME_PROT_READ ? 'r' : '-'),
+			 (e->kve_protection & KVME_PROT_WRITE ? 'w' : '-'),
+			 (e->kve_protection & KVME_PROT_EXEC ? 'x' : '-'),
+			 e->kve_path);
+		prof_dump_write(propagate_err, map);
+loop_cont:
+		low += esz;
+	}
+
+	ret = false;
```

Comment:
This still doesn't look right; the last branch `if (mfd != -1) { } else {} ` will do `ret = true;`. Let's either do `goto label_return`, or refactor so that mfd isn't part of the FreeBSD path.